common/utils.go

```diff
@@ -84,3 +90,18 @@ func IsNil(i interface{}) bool {
 	}
 	return false
 }
+
+func Go(f func(), panicHandlers ...func(interface{})) {
```

Comment:
1. `go foo()` is such a basic construction in Go. I don't think it's a good idea to wrap it into some `common.Go()` and force all developers to use it instead. Neither I think it's feasible.

2. We could instead create a `common.Recover` function that devs would call in their gorouties. This way we don't overcomplicate the code:
    ```go
    go Foo() {
        defer common.Recover()
        // Foo code
    }
    ```

@status-im/status-go-guild what do you think is better?

Reply:
I agree with Igor here. Wrapping common `go` idiom into `common.Go` will result in developers' headache. Moreover, I would try to be as specific as possible:
```golang
go func() {
    defer common.LogOnPanic()
    // Foo code
}
```

Do we need `panicHandlers ...func(interface{})` for anything other than tests?

Reply:
Instead of using `common`, wdyt about using the package name `safe`?  i.e.:  `safe.Go(someFn)`.

Reply:
> If we do this, we will lose the original intention of creating this PR. The purpose of creating this PR is to capture panics that occur in all goroutines and record logs for quick problem localization in any env and collect error logs on the production version if possible.

I am a bit confused, the proposed alternative should behave exactly the same. Instead of:
```diff
-        go func() {
+        utils.Go(func() {
             <-available
             m.InitHistoryArchiveTasks(controlledCommunities)
-        }()
+        })
```
it would be:
```diff
         go func() {
+            defer common.LogOnPanic()
             <-available
             m.InitHistoryArchiveTasks(controlledCommunities)
         }()
```

Reply:
> If we only add defer common.LogOnPanic() for specific goroutines, this cannot ensure that we can capture panics that occur in newly written goroutines in the future.

I see your point now. But isn't it essentially the same when developers forget to use `utils.Go`? I agree that it's easier to enforce the use of `utils.Go` across the codebase compared to defer `common.LogOnPanic()`, but the former comes with a bigger commitment and less flexibility.

- `utils.Go` will become a centralized abstraction that might not suit every use case. Would you use `utils.Go` or `go func(){}` in tests and tools, for instance?
- `utils.Go` is not obvious. Developers might not fully understand what it does internally. Additionally, there may be a temptation to add extra logic to `utils.Go` in the future, leading to incorrect assumptions based on its previous behavior.
- `utils.Go` might make it more difficult for linters to understand and analyze the code - linters are generally tuned to recognize common Go idioms.

Reply:
> If we use the first method (utils.Go), we can use the regular expression like go\s+([\w.]+)\(.*?\)|go\s+func\s*\(.*?\)\s*{ to detect whether the native way of creating goroutines is used in the code. If we use the second method (defer), implementing detection is not so easy?

That's true :disappointed: We could add checks but we'll certainly miss some scenarios.

- Check for invoking anonymous func is pretty simple
```
go func() {
   // code
}()
```

- Check for invoking defined func can be challenging but doable
```
go foo()
```

- Check for invoking func pointer is not doable
```
func runAsync(fn func()) {
  go fn()
}
```

Reply:
I made a quick I wish it was quicker though) test. The results are pretty impressing:

![image](https://github.com/user-attachments/assets/19f64fa3-6a6f-44ab-9d25-cdf66161b03f)

It's capable of checking anonymous and non-anonymous functions and methods. 
Still need to optimize the calls to `gopls definition` are taking 4-5 seconds each now. This is unacceptable, but doable to optimize and work as quick as your IDE.

The case that @osmaczko described is indeed very complex though, but it's pretty rare:
```go
func runAsync(fn func()) {
  go fn()
}
```

<details>
    <summary>Code for who's interested</summary>

```go
package main

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"

	"github.com/ethereum/go-ethereum/log"
)

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run main.go <directory>")
		return
	}

	// Initialize logger with colors
	handler := log.StreamHandler(os.Stdout, log.TerminalFormat(true))
	log.Root().SetHandler(log.LvlFilterHandler(log.LvlInfo, handler))

	dir := os.Args[1]
	log.Info("Starting analysis...", "directory", dir)

	// Step 1: Scan all files and look for `go` calls
	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			log.Error("Error walking the path", "path", dir, "error", err)
			return err
		}
		if info.IsDir() || !strings.HasSuffix(info.Name(), ".go") {
			return nil
		}

		log.Info("Scanning Go file", "file", path)
		checkFileForGoroutines(path)
		return nil
	})

	if err != nil {
		log.Error("Error during file walk", "error", err)
	}

	log.Info("Analysis complete")
}

// checkFileForGoroutines scans a Go file for any `go` statements (goroutines)
func checkFileForGoroutines(filePath string) {
	file, err := os.Open(filePath)
	if err != nil {
		log.Error("Error opening file", "file", filePath, "error", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var lineNumber int
	// Regex for non-anonymous function/method calls: `go functionName()`
	regex := regexp.MustCompile(`go\s+(\.|\w)+\(\)$`)

	for scanner.Scan() {
		lineNumber++
		line := scanner.Text() // Do not trim spaces here

		// Detect anonymous goroutines
		if strings.Contains(line, "go func") {
			log.Info("Found anonymous goroutine", "file", filePath, "line", lineNumber, "lineContent", line)
			checkAnonymousGoroutine(filePath, lineNumber)
			continue
		}

		// Detect non-anonymous goroutines using regex
		if regex.MatchString(line) {
			log.Info("Found non-anonymous goroutine", "file", filePath, "line", lineNumber, "lineContent", line)

			// Find the position of the first occurrence of "()"
			cursorPos := strings.Index(line, "()")
			if cursorPos == -1 {
				log.Error("Failed to find function call", "file", filePath, "line", lineNumber)
				continue
			}

			// Calculate the cursor position by adjusting for tabs (counting tabs as 4 characters)
			//tabs := strings.Count(line[:cursorPos], "\t")
			//adjustedCursorPos := cursorPos - (tabs * 4) // Subtract 3 for each tab since a tab counts as 4 chars

			checkNamedFunction(filePath, lineNumber, cursorPos)
		}
	}

	if err := scanner.Err(); err != nil {
		log.Error("Error reading file", "file", filePath, "error", err)
	}
}

// checkAnonymousGoroutine checks if an anonymous goroutine has `defer utils.LogOnPanic()`
func checkAnonymousGoroutine(filePath string, lineNumber int) {
	//log.Debug("Checking anonymous goroutine", "file", filePath, "line", lineNumber)

	// Open the file again and scan from the `go func` line onwards
	file, err := os.Open(filePath)
	if err != nil {
		log.Error("Error opening file", "file", filePath, "error", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var currentLine int
	for scanner.Scan() {
		currentLine++
		if currentLine <= lineNumber {
			continue
		}

		line := scanner.Text()
		// First line of the function body
		if strings.Contains(line, "defer utils.LogOnPanic()") {
			log.Info("Found defer utils.LogOnPanic() in anonymous function", "file", filePath, "line", lineNumber)
		} else {
			log.Warn("Missing defer utils.LogOnPanic() in anonymous function", "file", filePath, "line", lineNumber)
		}
		return
	}

	if err := scanner.Err(); err != nil {
		log.Error("Error reading file", "file", filePath, "error", err)
	}
}

// extractFunctionNameAndCursorPosition extracts the function or method name and calculates the cursor position
func extractFunctionNameAndCursorPosition(line string, matches []string) (string, int) {
	funcName := matches[1]

	// Calculate the cursor position (count tabs as one character)
	regex := regexp.MustCompile(`\bgo\s+`)
	cursorPos := regex.FindStringIndex(line)[1] // Position after `go ` keyword

	return funcName, cursorPos
}

// checkNamedFunction uses `gopls` to find the definition of a named function/method and checks its first line
func checkNamedFunction(filePath string, lineNumber, charPos int) {
	log.Debug("Checking named function", "file", filePath, "line", lineNumber, "char", charPos)

	// Use `gopls` to find the definition of the function/method
	cmd := exec.Command("gopls", "definition", fmt.Sprintf("%s:%d:%d", filePath, lineNumber, charPos))
	output, err := cmd.Output()
	if err != nil {
		log.Error("Error running gopls definition", "file", filePath, "line", lineNumber, "error", err)
		return
	}

	definitionOutput := string(output)
	// Parse the definition output to find the file and line number of the function definition
	parseAndCheckFunctionDefinition(definitionOutput)
}

// parseAndCheckFunctionDefinition parses the output of `gopls definition` and checks the function body
func parseAndCheckFunctionDefinition(definitionOutput string) {
	// The output of `gopls definition` will contain the file path and position of the function definition
	// Example output might be:
	// /path/to/file.go:23:5
	log.Debug("Parsed definition", "definition", definitionOutput)

	// Extract file path and line number from the output
	parts := strings.Split(definitionOutput, ":")
	if len(parts) < 2 {
		log.Error("Failed to parse gopls definition output", "output", definitionOutput)
		return
	}
	defFilePath := parts[0]
	lineNumber := atoi(parts[1])

	// Open the file and check the first statement inside the function body
	checkFirstLineInFunctionBody(defFilePath, lineNumber)
}

// checkFirstLineInFunctionBody checks the first line inside a function body for `defer utils.LogOnPanic()`
func checkFirstLineInFunctionBody(filePath string, startLine int) {
	log.Debug("Checking function body", "file", filePath, "startLine", startLine)

	file, err := os.Open(filePath)
	if err != nil {
		log.Error("Error opening file", "file", filePath, "error", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var currentLine int
	for scanner.Scan() {
		currentLine++
		if currentLine <= startLine {
			continue
		}

		line := scanner.Text()

		if strings.Contains(line, "defer utils.LogOnPanic()") {
			log.Info("Found defer utils.LogOnPanic() in function", "file", filePath, "line", startLine)
		} else {
			log.Warn("Missing defer utils.LogOnPanic() in function", "file", filePath, "line", startLine)
		}

		return
	}

	if err := scanner.Err(); err != nil {
		log.Error("Error reading file", "file", filePath, "error", err)
	}
}

// atoi is a helper to safely convert a string to an int
func atoi(s string) int {
	i, err := strconv.Atoi(s)
	if err != nil {
		return 0
	}
	return i
}
```
</details>

Reply:
I agree that catching `go anythiing` is much-much simpler. And it will do 100% of cases.
But forcing devs to use `safe.Go` over default syntax sounds very not good to me.